ledger/simulation/simulation_eval_test.go

```diff
@@ -4353,136 +4347,1748 @@ int 1`,
+		GlobalInitialStatesTestCase{
+			prepareInstruction: [][][]byte{
+				{
+					[]byte("A"), []byte("initial content A"),
+				},
+			},
+			txnsArgs: [][][]byte{
+				{
+					[]byte("put"), []byte("A"), []byte("content A"),
+				},
+				{
+					[]byte("del"), []byte("A"),
+				},
+			},
+			initialGlobalStates: simulation.AppKVPairs{
+				"A": basics.TealValue{
+					Type:  basics.TealBytesType,
+					Bytes: "initial content A",
+				},
+			},
+		},
+	)
+}
+
+type LocalStateOperation struct {
+	addressIndex uint64
+	appArgs      [][]byte
+}
+
+type LocalInitialStatesTestCase struct {
+	prepareInstructions  []LocalStateOperation
+	simulateInstructions []LocalStateOperation
+	initialLocalStates   map[uint64]simulation.AppKVPairs
+}
+
+func (testcase LocalInitialStatesTestCase) toSignedTxns(env simulationtesting.Environment, appID basics.AppIndex) []transactions.SignedTxn {
+	txns := make([]*txntest.Txn, len(testcase.simulateInstructions))
+	for i, instruction := range testcase.simulateInstructions {
+		tempTxn := env.TxnInfo.NewTxn(txntest.Txn{
+			Type:            protocol.ApplicationCallTx,
+			Sender:          env.Accounts[instruction.addressIndex].Addr,
+			ApplicationID:   appID,
+			ApplicationArgs: instruction.appArgs,
+		})
+		txns[i] = &tempTxn
+	}
+	txntest.Group(txns...)
+	signedTxns := make([]transactions.SignedTxn, len(testcase.simulateInstructions))
+	for i, txn := range txns {
+		signedTxns[i] = txn.Txn().Sign(env.Accounts[testcase.simulateInstructions[i].addressIndex].Sk)
+	}
+	return signedTxns
+}
+
+func testLocalInitialStatesHelper(t *testing.T, testcase LocalInitialStatesTestCase) {
+	t.Helper()
+
+	myEnv := simulationtesting.PrepareSimulatorTest(t)
+	defer myEnv.Close()
+
+	proto := myEnv.TxnInfo.CurrentProtocolParams()
+	appCreator := myEnv.Accounts[0]
+
+	approvalProgramSrc := `#pragma version 8
+txn ApplicationID
+bz end // Do nothing during create
+
+txn OnCompletion
+int OptIn
+==
+bnz end // Always allow optin
+
+byte "put"
+byte "get"
+byte "del"
+
+txn ApplicationArgs 0
+match put get del
+err // Unknown command
+
+put:
+  txn Sender            // account
+  txn ApplicationArgs 1 // key
+  txn ApplicationArgs 2 // local state content
+  app_local_put
+  b end
+
+get:
+  txn Sender            // account
+  txn ApplicationArgs 1 // key
+  app_local_get
+  pop
+  b end
+
+del:
+  txn Sender            // account
+  txn ApplicationArgs 1 // key
+  app_local_del
+  b end
+
+end:
+  int 1
+`
+
+	appID := myEnv.CreateApp(appCreator.Addr, simulationtesting.AppParams{
+		LocalStateSchema: basics.StateSchema{NumByteSlice: 8},
+		ApprovalProgram:  approvalProgramSrc,
+		ClearStateProgram: `#pragma version 8
+int 1`,
+	})
+
+	op, err := logic.AssembleString(approvalProgramSrc)
+	require.NoError(t, err)
+	progHash := crypto.Hash(op.Program)
+
+	transferable := myEnv.Accounts[1].AcctData.MicroAlgos.Raw - proto.MinBalance - proto.MinTxnFee
+	myEnv.TransferAlgos(myEnv.Accounts[1].Addr, appID.Address(), transferable)
```

Comment:
This step should not be necessary if you're not using boxes

Reply:
see [remove some unnecessary steps in testing](https://github.com/algorand/go-algorand/pull/5686/commits/20cb6c7d5c8777e46c9d530b22448a59e979eafb), good point.